lab_1/main.py

```diff
@@ -4,68 +4,135 @@
 """
 
 
-def tokenize(text: str) -> list or None:
+import json
+
+from os.path import exists
+from typing import Optional
+
+
+def tokenize(text: str) -> Optional[list]:
     """
     Splits a text into tokens, converts the tokens into lowercase,
     removes punctuation and other symbols from words
     :param text: a text
     :return: a list of lower-cased tokens without punctuation
     """
-    pass
+    if isinstance(text, str):
+        tokenized = (
+            "".join([char for char in text if char.isalpha() or char.isspace()])
+            .strip()
+            .lower()
+            .split()
+        )
+        return tokenized
+    return None
 
 
-def remove_stop_words(tokens: list, stop_words: list) -> list or None:
+def remove_stop_words(tokens: list, stop_words: list) -> Optional[list]:
     """
     Removes stop words
     :param tokens: a list of tokens
     :param stop_words: a list of stop words
     :return: a list of tokens without stop words
     """
-    pass
+    if (
+        isinstance(tokens, list)
+        and isinstance(stop_words, list)
+        and all(isinstance(i, str) for i in tokens)
+    ):
+        cleaned = [word for word in tokens if word not in stop_words]
+        return cleaned
+    return None
 
 
-def calculate_frequencies(tokens: list) -> dict or None:
+def calculate_frequencies(tokens: list) -> Optional[dict]:
     """
     Calculates frequencies of given tokens
     :param tokens: a list of tokens
     :return: a dictionary with frequencies
     """
-    pass
+    if isinstance(tokens, list) and all(isinstance(i, str) for i in tokens):
+        freq_dict = {}
+        for word in tokens:
+            freq_dict[word] = tokens.count(word)
+        return freq_dict
+    return None
 
 
-def get_top_n_words(freq_dict: dict, top_n: int) -> list or None:
+def get_top_n_words(freq_dict: dict, top_n: int) -> Optional[list]:
     """
     Returns the most common words
     :param freq_dict: a dictionary with frequencies
     :param top_n: a number of the most common words
     :return: a list of the most common words
     """
-    pass
+    if (
+        isinstance(freq_dict, dict)
+        and isinstance(top_n, int)
+        and all(isinstance(value, int) for value in freq_dict.values())
+    ):
+        sorted_dict = [
```

Comment:
do you like this line breaking?
seems a bit hard to read

pep8 allows several ways

Reply:
Actually, I do. But if you don't - I'll change the design.

Reply:
no prob then